src/JBSwapTerminal.sol

```diff
@@ -0,0 +1,535 @@
+
+    /// @notice
+    ///  _projectId The ID of the project being paid.
+    ///  _amount The amount of terminal tokens being received, as a fixed point number with the same amount of decimals
+    /// as this terminal. If this terminal's token is ETH, this is ignored and msg.value is used in its place.
+    ///  _token The token being paid. This terminal ignores this property since it only manages one token.
+    ///  _beneficiary The address to mint tokens for and pass along to the funding cycle's data source and delegate.
+    ///  _minReturnedTokens The minimum number of project tokens expected in return, as a fixed point number with the
+    /// same amount of decimals as this terminal.
+    ///  _memo A memo to pass along to the emitted event.
+    ///  _metadata Bytes to send along to the data source, delegate, and emitted event, if provided.
+    /// @return The number of tokens minted for the beneficiary, as a fixed point number with 18 decimals.
+    function pay(
+        uint256 _projectId,
+        address _token,
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _minReturnedTokens,
+        string calldata _memo,
+        bytes calldata _metadata
+    )
+        external
+        payable
+        virtual
+        override
+        returns (uint256)
+    {
+        SwapParams memory _swapParams;
+        _swapParams.projectId = _projectId;
+        _swapParams.tokenIn = _token;
+        _swapParams.amountIn = _amount;
+
+        {
+            // Check for a quote passed by the user
+            (bool _exists, bytes memory _parsedData) = JBMetadataResolver.getDataFor(bytes4("SWAP"), _metadata);
+
+            // If there is a quote, use it
+            if (_exists) {
+                (_swapParams.minAmountOut, _swapParams.pool, _swapParams.tokenOut) =
+                    abi.decode(_parsedData, (uint256, IUniswapV3Pool, address));
+                // If no quote, check there is a default pool assigned and get a twap
+            } else {
+                IUniswapV3Pool _pool = _poolFor[_projectId][_token][address(0)];
+
+                _swapParams.pool = _pool;
+
+                // If no default pool, revert - TODO: send back to the caller instead (could be either fee or EOA tho)
+                if (address(_pool) == address(0)) revert NO_DEFAULT_POOL_DEFINED();
+
+                (address _poolToken0, address _poolToken1) = (_pool.token0(), _pool.token1());
+                _swapParams.tokenOut = _poolToken0 == _token ? _poolToken1 : _poolToken0;
+
+                // Get a twap from the pool, includes a default max slippage
+                _swapParams.minAmountOut = _getTwapFrom(_swapParams);
+            }
+        }
+
+        IJBTerminal _terminal = DIRECTORY.primaryTerminalOf(_projectId, _token);
+
+        // Check the primary terminal accepts the token (save swap gas if not accepted)
+        if (address(_terminal) == address(0)) revert TOKEN_NOT_ACCEPTED(); // TODO: no revert?
```

Comment:
isn't the primary terminal of _projectId, _token potentially this swap terminal? we looking for the primary terminal of _swapParams.tokenOut?

Reply:
True!